cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformExecutor.java

```diff
@@ -68,73 +56,71 @@ public TransformResponse runOneIteration(IN input) throws Exception {
       }
     }
 
+
     Map<String, List<Object>> terminalNodeEntriesMap = new HashMap<>();
-    Map<String, List<Object>> emitterEntries = defaultEmitter.getEntriesMap();
-    for (String key : emitterEntries.keySet()) {
-      if (!connectionsMap.containsKey(key)) {
+    for (Map.Entry<String, TransformDetail> transformDetailEntry : transformDetailMap.entrySet()) {
```

Comment:
you can find the terminal nodes in the constructor rather than finding them for each iteration.